### PR TITLE
Feature/compliance

### PR DESCRIPTION
**Pull request by EvgenKor**
These changes are required to generate detailed compliance report within the client library. Added following routines in the wasm library:
- `extract_decrypt_keys` to get symmetric cipher keys needed to decrypt output tx items (account and notes)
- `symcipher_decrypt_acc` \ `symcipher_decrypt_acc` to decrypt account and notes from the memoblock (needed to verify report correctness during development)
- `get_tx_inputs` to find transaction inputs for the transaction: account and notes and intermediate nullifier from the input account
- `calculate_nullifier` to getting nullifier from the account
- few helper routines and unit-test for retrieving tx inputs

### CODE REVIEW

File: libzkbob-rs-wasm/src/client/mod.rs [hunk 1]
```diff
@@ -474,6 +496,41 @@ impl UserAccount {
         serde_wasm_bindgen::to_value(&data).unwrap()
     }
 
+    #[wasm_bindgen(js_name = "getTxInputs")]
+    /// Returns transaction inputs: account and notes
+    pub fn get_tx_inputs(&self, index: u64) -> Result<TxInput, JsValue> {
+        if index & constants::OUTPLUSONELOG as u64 != 0 {
```
Comment (AllFi) [inline libzkbob-rs-wasm/src/client/mod.rs:502]: This check doesn't look correct. For example 8 & 7 == 0 too but it doesn't mean that 8 % 128 == 0. I guess we should use constants::OUT instead but, as for me, index % (constants::OUT + 1) == 0 looks much better even if it's a bit slower.

File: libzkbob-rs/src/client/state.rs [hunk 2]
```diff
@@ -175,6 +175,40 @@ where
...
+    pub fn get_previous_account(&self, index: u64) -> Option<(u64, Account<P::Fr>)> {
+        if index == 0 { return None }
+
+        let prev_acc_indexes = self.txs
+            .iter_slice(0..=(index-1))
+            .filter_map(|(idx, tx)| match tx {
+                Transaction::Account(_) => Some(idx),
+                _ => None,
+            })
+            .max();
```
Comment (AllFi) [inline libzkbob-rs/src/client/state.rs:194]: Maybe it will look better this way:
```
self.txs
  .iter_slice(0..=(index-1))
  .filter_map(|(idx, tx)| match tx {
      Transaction::Account(acc) => Some((idx, acc)),
      _ => None,
  })
  .last()
```

File: libzkbob-rs/src/client/state.rs [hunk 3]
```diff
@@ -175,6 +175,40 @@ where
...
+
+        match prev_acc_indexes {
+            Some(idx) => Some((idx, self.get_account(idx).unwrap())),
+            _ => None,
+        }
+    }
+
+    pub fn get_notes_in_range(&self, range: Range<u64>) -> Vec<(u64, Note<P::Fr>)> {
+        self.txs
+            .iter_slice(range.start..=(if range.end==0 { 0 } else { range.end-1 }))
```
Comment (AllFi) [inline libzkbob-rs/src/client/state.rs:204]: I guess we can replace it with `range.end.saturating_sub(1)`. Wdyt?

**Review by AllFi — CHANGES_REQUESTED** [review]

File: libzkbob-rs/src/client/mod.rs [hunk 4]
```diff
@@ -69,6 +71,16 @@ pub struct TransactionData<Fr: PrimeField> {
     pub out_hashes: SizedVec<Num<Fr>, { constants::OUT + 1 }>,
 }
 
+#[derive(Serialize, Deserialize, Clone)]
+pub struct TransactionInputs<Fr: PrimeField> {
+    #[serde(bound(serialize = "", deserialize = ""))]
```
Comment (r0wdy1) [inline libzkbob-rs/src/client/mod.rs:76]: is it really needed? I see no problem with default serde behavior  here

File: libzkbob-rs-wasm/src/client/mod.rs (hunk 1 above)
Comment (EvgenKor) [reply libzkbob-rs-wasm/src/client/mod.rs]: It's definitely a mistake made by me. Of course the `constants::OUT` should be here instead `constants::OUTPLUSONELOG`. The speed doesn't make sense here because it isn't time-critical operation. But I believe the error message in the next line leaves no way to doubt about check semantic

Fixed in https://github.com/zkBob/libzkbob-rs/pull/46/commits/c21ea801b890216200e3ed539fa6ea13e707f70c

File: libzkbob-rs/src/client/state.rs (hunk 2 above)
Comment (EvgenKor) [reply libzkbob-rs/src/client/state.rs:194]: Do not agree. The SparseArray object `txs` in the common case doesn't guarantee the index order. So the last element isn't 'account with the greatest index'. We are fill it in the serial manner on the top-level. But at this level we cannot make such assumptions

File: libzkbob-rs/src/client/state.rs (hunk 3 above)
Comment (EvgenKor) [reply libzkbob-rs/src/client/state.rs]: Thanks, it looks more better

Fixed in https://github.com/zkBob/libzkbob-rs/pull/46/commits/c21ea801b890216200e3ed539fa6ea13e707f70c

File: libzkbob-rs/src/client/mod.rs (hunk 4 above)
Comment (EvgenKor) [reply libzkbob-rs/src/client/mod.rs]: Most likely it's a copy-paste issue. Removed in https://github.com/zkBob/libzkbob-rs/pull/46/commits/c21ea801b890216200e3ed539fa6ea13e707f70c

File: libzkbob-rs/src/client/state.rs (hunk 2 above)
Comment (AllFi) [reply libzkbob-rs/src/client/state.rs:194]: Yep, it makes sense.

**Review by AllFi — APPROVED** [review]

**Comment by EvgenKor** [comment]
The PR is ready to review. The branch was merged with dd branch to resolve conflicts

**Review by AllFi — APPROVED** [review]

File: libzkbob-rs/Cargo.toml [hunk 5]
```diff
@@ -24,7 +24,7 @@ hex = { version = "0.4.3", features = ["serde"] }
 
 [dependencies.libzeropool]
 git = "https://github.com/zkbob/libzeropool-zkbob"
-branch = "master"
+branch = "feature/compliance"
```
Comment (r0wdy1) [inline libzkbob-rs/Cargo.toml:27]: may be revert this change?
 there seems to be nothing new

File: libzkbob-rs-wasm/src/client/mod.rs [hunk 6]
```diff
@@ -293,7 +312,8 @@ impl UserAccount {
 
     #[wasm_bindgen(js_name = "createTransferOptimistic")]
     pub fn create_tranfer_optimistic(&self, transfer: ITransferData, new_state: &JsValue) -> Result<Promise, JsValue> {
-        let new_state: StateUpdate = new_state.into_serde().map_err(|err| js_err!(&err.to_string()))?;
+        //let new_state: StateUpdate = new_state.into_serde().map_err(|err| js_err!(&err.to_string()))?;
```
Comment (AllFi) [inline libzkbob-rs-wasm/src/client/mod.rs:315]:
```suggestion
```

File: libzkbob-rs-wasm/src/client/mod.rs [hunk 7]
```diff
@@ -304,7 +324,8 @@ impl UserAccount {
 
     #[wasm_bindgen(js_name = "createWithdrawalOptimistic")]
     pub fn create_withdraw_optimistic(&self, withdraw: IWithdrawData, new_state: &JsValue) -> Result<Promise, JsValue> {
-        let new_state: StateUpdate = new_state.into_serde().map_err(|err| js_err!(&err.to_string()))?;
+        //let new_state: StateUpdate = new_state.into_serde().map_err(|err| js_err!(&err.to_string()))?;
```
Comment (AllFi) [inline libzkbob-rs-wasm/src/client/mod.rs:327]:
```suggestion
```

File: libzkbob-rs-wasm/src/client/mod.rs [hunk 8]
```diff
@@ -386,9 +407,11 @@ impl UserAccount {
 
     #[wasm_bindgen(js_name = "updateState")]
     pub fn update_state(&mut self, state_update: &JsValue, siblings: Option<TreeNodes>) -> Result<(), JsValue> {
-        let state_update: StateUpdate = state_update.into_serde().map_err(|err| js_err!(&err.to_string()))?;
+        //let state_update: StateUpdate = state_update.into_serde().map_err(|err| js_err!(&err.to_string()))?;
```
Comment (AllFi) [inline libzkbob-rs-wasm/src/client/mod.rs:410]:
```suggestion
```

File: libzkbob-rs-wasm/src/client/mod.rs [hunk 9]
```diff
@@ -386,9 +407,11 @@ impl UserAccount {
...
+        let state_update: StateUpdate = serde_wasm_bindgen::from_value(state_update.to_owned()).map_err(|err| js_err!(&err.to_string()))?;
         let siblings: Option<Vec<Node<Fr>>> = match siblings {
-            Some(val) => val.into_serde().map_err(|err| js_err!(&err.to_string()))?,
+            //Some(val) => val.into_serde().map_err(|err| js_err!(&err.to_string()))?,
```
Comment (AllFi) [inline libzkbob-rs-wasm/src/client/mod.rs:413]:
```suggestion
```

File: libzkbob-rs-wasm/src/client/tx_parser.rs [hunk 10]
```diff
@@ -92,7 +129,8 @@ impl TxParser {
         let params = &self.params;
         let eta = Keys::derive(sk, params).eta;
 
-        let txs: Vec<IndexedTx> = txs.into_serde().map_err(|err| js_err!(&err.to_string()))?;
+        //let txs: Vec<IndexedTx> = txs.into_serde().map_err(|err| js_err!(&err.to_string()))?;
```
Comment (AllFi) [inline libzkbob-rs-wasm/src/client/tx_parser.rs:132]:
```suggestion
```

Comment (AllFi) [inline libzkbob-rs-wasm/src/client/tx_parser.rs:132]:
```suggestion
```

**Review by AllFi — COMMENTED** [review]
I believe it's better to remove commented code in order to keep the code clean. Overall, it looks good to me.

**Review by AllFi — APPROVED** [review]
I believe it's better to remove commented code in order to keep the code clean. Overall, it looks good to me.

File: libzkbob-rs/Cargo.toml (hunk 5 above)
Comment (EvgenKor) [reply libzkbob-rs/Cargo.toml]: Reverted in https://github.com/zkBob/libzkbob-rs/pull/46/commits/b0a5549b05eaa4878be2cb8a8de349d90e61c970
